source/slang/slang-ir-inline.cpp

```diff
@@ -0,0 +1,451 @@
+                env.mapOldValToNew.Add(param, arg);
+            }
+            SLANG_ASSERT(argCounter == (Int)call->getArgCount());
+        }
+
+        // For now, our inlining pass only handles the case where
+        // the callee is a "trivial" function, which can support
+        // a very simple approach to inlining.
+        //
+        if( isTrivialFunc(callee) )
+        {
+            inlineTrivialFuncBody(callSite, &env, &builder);
+        }
+        else
+        {
+            // Running into any non-trivial function to be inlined
+            // is currently an internal compiler error.
+            //
+            SLANG_UNIMPLEMENTED_X("general case of inlining");
+        }
+    }
+
+        /// Check if `func` represents a trivial single-block callee that can be inlined simply
+    bool isTrivialFunc(IRFunc* func)
+    {
+        // The function must have a single bocy block to be trivial.
+        //
+        auto firstBlock = func->getFirstBlock();
+        if( firstBlock->getNextBlock() )
+            return false;
+
+        // If the body block is decorated (for some reason), then the function is non-trivial.
+        //
+        if( firstBlock->getFirstDecoration() )
+            return false;
+
+        // If the body block terminates in something other than a `return` then the function is non-trivial.
+        //
+        auto terminator = firstBlock->getTerminator();
+        if( !as<IRReturn>(terminator) )
+            return false;
+
+        return true;
+    }
+
+        /// Inline the body of the callee for `callSite`, where the callee is trivial as tested by `isTrivialFunc`
+    void inlineTrivialFuncBody(CallSiteInfo const& callSite, IRCloneEnv* env, IRBuilder* builder)
+    {
+        auto call = callSite.call;
+        auto callee = callSite.callee;
+        auto firstBlock = callee->getFirstBlock();
+
+        // We know that the callee has a single block, so if we encounter
+        // a `returnVal` instruction then it must be the one and only
+        // return point for the block, and its operand will be the value
+        // the calee returns.
+        //
+        IRInst* returnedValue = nullptr;
+
+        // We will loop over the instructions of the one and only block,
+        // and clone each of them appropriately.
+        //
+        for( auto inst : firstBlock->getChildren() )
+        {
+            switch( inst->op )
+            {
+            default:
+                // The default value is to clone the instruction using
+                // the existing cloning infrastructure and the `env`
+                // we have already set up.
+                //
+                cloneInst(env, builder, inst);
+                break;
+
+            case kIROp_Param:
+                // Parameters can be completely ignored in the single-block
+                // case, because they have all been replaced via `env`.
+                break;
+
+            case kIROp_ReturnVoid:
```

Comment:
I'm assuming that there can't be a ReturnVoid currently because of the constraint for 'trivialFunction' - so assume this is here for a later more general inlining.

Reply:
The "trivial function" check actually allows for any `IRReturn` (which is the base type for `IRReturnVoid` and `IRReturnVal`). There is no reason not to support inlining of a trivial function with a `void` return value, since in practice it is no different from inlining a function that returns an `int` where the caller ignores the return value.

My hope/goal is that in a future cleanup pass we actually eliminate `IRReturnVoid` and just have a single `IRReturn` that matches the current `IRReturnVal` case. Then a `void` return will just be a `returnVal` that happens to return a "void literal." Doing that would simplify code when handling functions because then every function returns *something* (even if it is a `void` value) and we no longer have to handle two different flavors of `return`s.